### ISSUE TITLE
Implement SCSS Compiling GIP

### ISSUE THREAD

**Issue by mkosowsk** [gitcoinco/GIPs#6]
The following is taken from https://github.com/gitcoinco/web/issues/560 but removes the Refactoring Step of "One-by-one refactor and lint those files so that they utilize common & reusable components and variables"

Per @olafghanizadeh, @razordaze, @mkosowsk, @e-n-l, @olafghanizadeh, et al

As I Gitcoin developer
I would like to have have consolidated styling (view layer abstracted away from logic)
So that I can quickly make global styling changes across the app and have re-useable components, variables etc. 

### Background

Current styling in the site is “[is messy and a hassle to edit](https://github.com/gitcoinco/web/issues/135)”
0. The CSS of gitcoin.co is a bit messy and a hassle to edit.
1. Templates that aren’t reusable across [Dashboard](https://github.com/gitcoinco/web/tree/master/app/dashboard), [Marketing](https://github.com/gitcoinco/web/tree/master/app/marketing), [Retail](https://github.com/gitcoinco/web/tree/master/app/retail), etc.
2. Non reusable variables and components so that it’s difficult to maintain consistency across the site

### Acceptance Criteria
0. Build process that leverages SCSS to consolidate styling across the app
1. Preserve all current functionality (do no harm)

### Technical Details:
base.css
+ explorer.css
+ retail.css

#### Refactoring Steps:
Add SCSS compiling and linting to the Webpack tasks
Utilize a package manager to include common libraries used (‘bootstrap, animate.css, etc.’)
Start by defining common variables (colors, fonts, forms) - @PixelantDesign
Include all current CSS files into the build

**Comment by owocki** [comment]
seems positive directionally to me

**Comment by owocki** [comment]
lets call this GIP-2 -- want to PR it in to the `gips/` dir?

**Comment by mkosowsk** [comment]
I would like to wait just a bit to make sure @olafghanizadeh and @e-n-l have had their say on the topic before a PR, pinging them now on Slack 👍

**Comment by e-n-l** [comment]
> Move the CSS and static assets in a separate place as a static location.

I believe this implies migration into a different repo, e.g., [creative](https://github.com/gitcoinco/creative)  or something new ("static"?) but it's not explicitly addressed, nor have we really clarified how we want that to work. Perhaps there's latitude for this in the resolution of this proposal?

I seem to recall that we were going to split that action item into a separate issue: If that's the case, then my understanding is that the static / compiled assets would be generated on / prior to build (i.e., not in version control at all) from versioned source assets. (images, javascript, sass, etc.) In this case, we'd require developing a build process to accommodate the new work flow; it would be imperative to have that in place **prior** to beginning work on this issue. 

Using a separate repo for source assets is a fast workaround: Leave Web as-is, develop SCSS / source assets in this other repo, build, and then (at least for now, as MVP) commit the built assets to Web. Eventually, you can merge that separate repo into Web after the static build process is tested / bulletproof, or setup a build pipeline for static that would deploy it to a CDN or fixed location that Web would reference.

**Comment by e-n-l** [comment]
Extending this idea:
> Using a separate repo for source assets is a fast workaround: Leave Web as-is, develop SCSS / source assets in this other repo, build, and then (at least for now, as MVP) commit the built assets to Web. Eventually, you can merge that separate repo into Web after the static build process is tested / bulletproof, or setup a build pipeline for static that would deploy it to a CDN or fixed location that Web would reference.

0. **Add intermediary / dev style folder(s) to [web](https://github.com/gitcoinco/web) for active development / WIP**
1. Assets in [web](https://github.com/gitcoinco/web)'s working folder(s) would form a backlog to integrate into the static asset source & SCSS ("S.A.S.S.", LOL) repo, thus avoiding merge conflicts in [web](https://github.com/gitcoinco/web) static assets.


Background convo:
@owocki: 
> i’m happy to bounty this.. i dont think its really in analysis any longer..
>
>my one worry is bountying it and then having it take like a month to migrate and having a ton of merge conflicts

elaniobro: 
> yea, I hear you. Maybe it makes sense to pick a single page, (probably homepage) and re-write it using .sass/.less Create the aformentioned methdologies, and then migrate page/section 1 at a time… ¯\\_(ツ)_/¯

@e-n-l:
> Do the static assets / CSS get messed with often in code commits? 

@mbeacom 
> It's rather frequent.

@e-n-l:
> what about having an intermediary style area, for things that are in development / new, then moving them to a backlog to incorporate into a more static main style repo?
> i.e., ship your code / features with MVP css, (or sass) and get that migrated into the main static source repo offline.

**Comment by joshmobley** [comment]
Having gone through this process of modularizing styles on an existing project, one approach I've used that worked about as well as any, was to create the styles/markup for components in isolation to the rest of the project -- as part of the documentation process (https://fractal.build/ as an example).

Then, once we're happy with the starting point for the library, we can begin rolling that markup out section by section to accomplish:

1. not having massive PRs
2. not dealing with long timelines and piles of conflicts

Additionally, using a documentation system like Fractal, or at least a similar approach, would eventually allow you to merge it into the project itself, auto-generating documentation and examples from the stylesheets themselves.

**Comment by Elaniobro** [comment]
To elaborate on my comments quotes above by @e-n-l, to migrate the entire view code base `.css/.less/.sass` it and will be a tremendous effort. Doing so should be highly organized and planned, *only* after a plan, methodology and philosophy has been discussed with the internal `gitcoin` team and feedback from the community has been given.

*B* lock *E* lement *M* odifier aka BEM is a great way to write both class names and css. There are many articles on the subject, so I will not link to any of them here. I will say, doing so allows for highly scalable `css` that can also be both extended and overwritten with ease. The use of `id` attributes on an element are also eliminated to help with what was mentioned above.

Further more, writing BEM methodology leads to the ability to write and create patterns, which in turn will result in more scalability, reusability and easier maintenance.

take for example this simple `<img>` element:

`<img src="https://gitcoin.co/static/v2/images/tldr/bounties.574eee54d651.jpg">`

using BEM it could be re-written in the following manner where `p` could stand for pattern. In this case `p-image` is the pattern block `__image` is the pattern element and `--jpg` is the modifier. 

```
<div class="p-image">
  <img class="p-image__image p-image__image--jpg" src="https://gitcoin.co/static/v2/images/tldr/bounties.574eee54d651.jpg">`
</div>
```

This lets us re-write .css from:

```
.tldr_container img {
    max-width: 250px;
    max-height: 250px;
}
```

to:

```
.p-image {
  &__image{
    &--jpg{
      max-width: 250px;
      max-height: 250px;
    }
  }
}
```
While this does seem like more to write, it also keeps the .css contained. This can also let us create a pattern for reusability and increase performance by using [reference](https://css-tricks.com/reference-imports-in-less-are-kinda-cool/):

`@import (reference) '/path/to/patterns/image/image.less`

I would also argue, to feasibly do so, we should pick a single page to start with vs the entire site. I would recommend starting with the `homepage` and moving on from there.

As for organizing the various files we should also consider breaking them out so any new developer coming in can quickly identify and implement bug-fix, enhancements, new features etc.. 

```
/styles/base/colors.less
/styles/base/colors.less
/styles/base/colors.less
```

**Comment by owocki** [comment]
hi all 👋 

i move we 
1. merge this GIP into the main repo
2. break out the tasks above into bountied issues on gitcoinco/web

does that sounds like the right move to ya'll?

**Comment by mkosowsk** [comment]
@owocki makes sense to me! 👍 Do you want to take point on merging this GIP into the main repo?

Per @Elaniobro I also think it's a good idea to do a homepage issue and see how it plays out and then iterate from there 👍

**Comment by mbeacom** [comment]
Sounds like the right move to me!
I agree with @mkosowsk and @Elaniobro regarding chunking out this initiative into smaller segments with staggered bounty schedule starting with the homepage.

**Comment by owocki** [comment]
> Do you want to take point on merging this GIP into the main repo?
yes

**Comment by mkosowsk** [comment]
@owocki I will do PR before EOD :) following format of https://github.com/gitcoinco/GIPs/blob/master/gip-X.md

Think we have all the info here just want to make sure to process it into the established format of Abstract, Motivation, Specification, Rationale, Backwards Compatibility, Test Cases, and Implementation 👍

**Comment by mkosowsk** [comment]
@owocki @mbeacom et al PR at https://github.com/gitcoinco/GIPs/pull/12 :)

**Comment by mkosowsk** [comment]
Is Webpack the road forward or should we look into something like Django.SCSS?

**Comment by owocki** [comment]
Considerations for [Django SCSS] vs Webpack:

* What does the front end dev who we hire want to do?
* What other architecture on the frontend will depend upon webpack?
* Is webpack overkill from a dependency standpoint?

**Comment by mkosowsk** [comment]
Other discussion was on doing something simpler like off chain bounties at gitcoin.co/universe before something bigger like the homepage

**Comment by joshmobley** [comment]
For the record, you can have multiple entry and output files w/ webpack (i.e. it doesn't force you to bundle everything into 1 output file) -- but I'm indifferent on the tooling at this point, as it doesn't sound like we'd necessarily be leveraging webpack for much more than compilation of styles.
